thrust/thrust/transform.h

```diff
@@ -725,6 +725,161 @@ ForwardIterator transform_if(
   BinaryFunction binary_op,
   Predicate pred);
 
+//! Like \ref transform, but uses an element count instead of an iterator to the last element of the input sequence.
+template <typename DerivedPolicy, typename InputIterator, typename Size, typename OutputIterator, typename UnaryFunction>
+_CCCL_HOST_DEVICE OutputIterator transform_n(
+  const detail::execution_policy_base<DerivedPolicy>& exec,
+  InputIterator first,
+  Size count,
+  OutputIterator result,
+  UnaryFunction op)
+{
+  _CCCL_NVTX_RANGE_SCOPE("thrust::transform_n");
+  return thrust::transform(exec, first, first + count, result, op);
+}
+
+//! Like \ref transform, but uses an element count instead of an iterator to the last element of the input sequence.
+template <typename InputIterator, typename Size, typename OutputIterator, typename UnaryFunction>
+OutputIterator transform_n(InputIterator first, Size count, OutputIterator result, UnaryFunction op)
+{
+  _CCCL_NVTX_RANGE_SCOPE("thrust::transform_n");
+  return thrust::transform(first, first + count, result, op);
+}
+
+//! Like \ref transform, but uses an element count instead of an iterator to the last element of the input sequence.
+template <typename DerivedPolicy,
+          typename InputIterator1,
+          typename Size,
+          typename InputIterator2,
+          typename OutputIterator,
+          typename BinaryFunction>
+_CCCL_HOST_DEVICE OutputIterator transform_n(
+  const detail::execution_policy_base<DerivedPolicy>& exec,
+  InputIterator1 first1,
+  Size count,
+  InputIterator2 first2,
+  OutputIterator result,
+  BinaryFunction op)
+{
+  _CCCL_NVTX_RANGE_SCOPE("thrust::transform_n");
+  return thrust::transform(exec, first1, first1 + count, first2, result, op);
+}
+
+//! Like \ref transform, but uses an element count instead of an iterator to the last element of the input sequence.
+template <typename InputIterator1, typename Size, typename InputIterator2, typename OutputIterator, typename BinaryFunction>
+OutputIterator
+transform_n(InputIterator1 first1, Size count, InputIterator2 first2, OutputIterator result, BinaryFunction op)
+{
+  _CCCL_NVTX_RANGE_SCOPE("thrust::transform_n");
+  return thrust::transform(first1, first1 + count, first2, result, op);
+}
+
+//! Like \ref transform_if, but uses an element count instead of an iterator to the last element of the input sequence.
+template <typename DerivedPolicy,
+          typename InputIterator,
+          typename Size,
+          typename ForwardIterator,
+          typename UnaryFunction,
+          typename Predicate>
+_CCCL_HOST_DEVICE ForwardIterator transform_if_n(
+  const detail::execution_policy_base<DerivedPolicy>& exec,
+  InputIterator first,
+  Size count,
+  ForwardIterator result,
+  UnaryFunction op,
+  Predicate pred)
+{
+  _CCCL_NVTX_RANGE_SCOPE("thrust::transform_if_n");
+  return thrust::transform(exec, first, first + count, result, op, pred);
```

Comment:
Nitpick: depending on the iterator type this could get us warnings because there is a conversion from `size_t` to `iter_difference<InputIterator>`

Should we rather take `iter_difference<InputIterator>` as the argument? An alternative would be to explicitly cast to it but then we should also check for overflow

Reply:
I was taking inspiration from [`copy_n`](https://en.cppreference.com/w/cpp/algorithm/copy_n), which has this signature:
```c++
template< class InputIt, class Size, class OutputIt >
OutputIt copy_n( InputIt first, Size count, OutputIt result );
```
You are correct that this can lead to a warning if there is a narrowing conversion.

> Should we rather take `iter_difference<InputIterator>` as the argument?

This is viable, but would result in the same warning. Just not in the Thrust header, but at the callee site. I am fine with this too.

> An alternative would be to explicitly cast to it but then we should also check for overflow

I would not try to suppress a warning that is potentially useful.

Reply:
This is something that was brought in by ranges, see https://en.cppreference.com/w/cpp/algorithm/ranges/copy_n

Reply:
That makes sense, let's switch to `iter_difference_t<InputIterator>` then